api/web3server.go

```diff
@@ -401,7 +401,20 @@ func (svr *web3Handler) getBalance(in *gjson.Result) (interface{}, error) {
 	if err != nil {
 		return nil, err
 	}
-	accountMeta, _, err := svr.coreService.Account(ioAddr)
+	bnParam := in.Get("params.1")
+	bn, err := parseBlockNumber(&bnParam)
+	if err != nil {
+		return nil, err
+	}
+	var (
+		accountMeta     *iotextypes.AccountMeta
+		height, archive = blockNumberToHeight(bn)
+	)
+	if !archive {
+		accountMeta, _, err = svr.coreService.Account(ioAddr)
+	} else {
+		accountMeta, _, err = svr.coreService.WithHeight(height).Account(ioAddr)
+	}
```

Comment:
such that we can always return a `Reader` from `svr.coreService` by calling `svr.coreService.Reader(bn).Account(ioAddr)`